### PR TITLE
Add environment variable to force flattening of 3D input tensor

### PR DESCRIPTION
**Pull request by kwen2501**
Adding an environment variable `TORCH_LINEAR_FLATTEN_3D` to force flattening of 3D input tensor even when it is non-contiguous.

Today, the `Linear` op would flatten a 3D input sensor if it is contiguous.

It was found that even for some non-contiguous inputs (esp. with BF16 data type), flattening would also yield higher performance.
For example:
```
x_size = (3072, 1196, 128)
x = torch.rand(x_size, device="cuda", dtype=torch.bfloat16)
x = torch.transpose(x, 1, 2)
torch._C._nn.linear(x, weight, bias)
```

Since the detailed auto-tuning is unknown, this PR adds an environment variable for users to make a choice.
(Default value is still 0.)

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -38,6 +38,28 @@
...
+  static int value = -1;
+  if (value == -1) {
+    const char* env_str = std::getenv("LINEAR_FLATTEN_3D");
+    if (env_str != nullptr && strcmp(env_str, "1") == 0) {
+      value = 1;
+    } else {
+      value = 0;
+    }
+  }
+  return value;
```
Comment (ngimel) [inline aten/src/ATen/native/Linear.cpp:52]: nit: return bool(value)

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -38,6 +38,28 @@
...
+}
+
+// `_flatten_3d_linear` flattens the first two dimensions of the input tensor
+// before passing it to linear operation if it is 3D
+static inline Tensor _flatten_3d_linear(const Tensor& input, const Tensor& weight, const Tensor& bias) {
+    const auto input_sizes = input.sym_sizes();
+    const auto result = at::addmm(bias, input.view_symint({input_sizes[0] * input_sizes[1], input_sizes[2]}), weight.t());
```
Comment (zhaojuanmao) [inline aten/src/ATen/native/Linear.cpp:59]: *bias?

Comment (kwen2501) [reply aten/src/ATen/native/Linear.cpp:60]: Used `*bias` when calling into `_flatten_3d_linear`

**Pull request closed without merging**